dev/src/bulk-writer.ts

```diff
@@ -0,0 +1,428 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+import assert = require('assert');
+import {FieldPath, Firestore, Timestamp} from '.';
+import {logger} from './logger';
+import {DocumentReference} from './reference';
+import {DocumentData, Precondition, SetOptions, UpdateData} from './types';
+import {Deferred, requestTag} from './util';
+import {BatchWriteResult, WriteBatch, WriteResult} from './write-batch';
+
+/**
+ * The maximum number of writes that be in a single batch.
+ */
+const BULK_WRITER_MAX_BATCH_SIZE = 500;
+
+/**
+ * Used to represent a batch on the BatchQueue.
+ *
+ * @private
+ */
+interface BulkCommitBatch {
+  // The underlying WriteBatch used for making writes.
+  writeBatch: WriteBatch;
+
+  // The set of references present in the WriteBatch.
+  refsInBatch: Set<DocumentReference>;
+
+  // A unique identifier for the batch.
+  tag: string;
+
+  // Whether the batch has been marked to be sent. Writes cannot be added to
+  // pending batches.
+  pending: boolean;
+
+  // A deferred promise that will return the result of writing the batch.
+  result: Deferred<BatchWriteResult[]>;
+}
+
+/**
+ * A Firestore BulkWriter than can be used to perform large amounts of writes in
+ * parallel. Writes to the same document will be executed sequentially.
+ *
+ * @class
+ */
+export class BulkWriter {
+  /**
+   * A queue of batches to be written.
+   */
+  private batchQueue: BulkCommitBatch[] = [];
+
+  /**
+   * A mapping of each batch's unique tag to a promise that will resolve with
+   * the result of writing the batch.
+   */
+  private resultsMap = new Map<string, Deferred<BatchWriteResult[]>>();
+
+  /**
+   * A set of all document references that are currently in flight.
+   */
+  private refsInFlight = new Set<DocumentReference>();
+  constructor(
+    private readonly firestore: Firestore,
+    private readonly maxBatchSize = BULK_WRITER_MAX_BATCH_SIZE
+  ) {}
+
+  /**
+   * Create a document with the provided object values. This will fail the batch
```

Comment:
`with the provided data`?

I would try to copy as much as you can from the existing WriteBatch API documentation.